flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -474,7 +484,8 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
         DataInputStatus status = inputProcessor.processInput();
         switch (status) {
             case MORE_AVAILABLE:
-                if (recordWriter.isAvailable()) {
+                if (recordWriter.isAvailable()
+                        && changelogWriterAvailabilityProvider.isApproximatelyAvailable()) {
```

Comment:
Why `isApproximatelyAvailable` and not `isAvailable`? Isn't this actually an anti optimisation? If the `changelogWriterAvailabilityProvider.getAvailableFuture() != AVAILABLE`, it should be actually cheaper to check `changelogWriterAvailabilityProvider.getAvailableFuture().isDone()` rather then suspend mailbox and wake it up via `CompletableFuture#thenRun` callback.

Reply:
You're right, I'll change to `isAvailable`.